src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldNotContainUnderscores.Fixer.cs

```diff
@@ -1,29 +1,87 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Immutable;
+using System.Composition;
+using System.Globalization;
+using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Rename;
 
 namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
 {
     /// <summary>
     /// CA1707: Identifiers should not contain underscores
     /// </summary>
-    public abstract class IdentifiersShouldNotContainUnderscoresFixer : CodeFixProvider
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class IdentifiersShouldNotContainUnderscoresFixer : CodeFixProvider
     {
-        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray<string>.Empty;
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IdentifiersShouldNotContainUnderscoresAnalyzer.RuleId);
 
         public sealed override FixAllProvider GetFixAllProvider()
         {
             // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
             return WellKnownFixAllProviders.BatchFixer;
         }
 
-        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
         {
-            // Fixer not yet implemented.
-            return Task.CompletedTask;
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var model = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
 
+            var node = root.FindNode(context.Span);
+            if (node == null)
+            {
+                return;
+            }
+
+            var symbol = model.GetDeclaredSymbol(node, context.CancellationToken);
+            if (symbol == null)
+            {
+                return;
+            }
+
+            string title = MicrosoftCodeQualityAnalyzersResources.IdentifiersShouldNotContainUnderscoresFixTitle;
+            context.RegisterCodeFix(new MyCodeAction(title,
```

Comment:
IMO we should do the same as what I am trying to do for the remove `Opt` suffix PR and check that the new symbol name does not already exists before registering the code fix. See https://github.com/dotnet/roslyn-analyzers/pull/3352/files#diff-6fee323e5832f875afec52f98a70ba86R48

Reply:
Great point! @Evangelink.
Symbol name clashing didn't come to my mind 😄 
I'll work on that.

Reply:
If I am not mistaken this comment is now resolved in line 56, @Youssef1313 could you confirm?